File: exercise/23/solution1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* merge2Lists(ListNode* list1, ListNode* list2) {
        // a fake head
        ListNode head(0);
        ListNode *tail = &head;
        while (list1 && list2) {
            if (list1->val < list2->val) {
                tail->next = list1;
                list1 = list1->next;
                tail = tail->next;
            } else {
                tail->next = list2;
                list2 = list2->next;
                tail = tail->next;
            }
        }
        if (list1) {
            tail->next = list1;
        } else {
            tail->next = list2;
        }
        return head.next;
    }

    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if (lists.empty()) {
            return NULL;
        }
        for (int gap = 1; gap < lists.size(); gap *= 2) {
            for (int k = 0; k + gap < lists.size(); k += gap * 2) {
                lists[k] = merge2Lists(lists[k], lists[k+gap]);
            }
        }
        return lists[0];
    }
};

File: exercise/23/solution1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def mergeTwoLists(self, list_a: ListNode, list_b: ListNode) -> ListNode:
        head = ListNode(0)
        tail = head
        while list_a and list_b:
            if list_a.val < list_b.val:
                tail.next = list_a
                list_a = list_a.next
            else:
                tail.next = list_b
                list_b = list_b.next
            tail = tail.next
        if list_a:
            tail.next = list_a
        else:
            tail.next = list_b
        return head.next

    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        if len(lists) == 0:
            return None
        gap = 1
        while gap < len(lists):
            for i in range(0, len(lists) - gap, gap * 2):
                lists[i] = self.mergeTwoLists(lists[i], lists[i+gap])
            gap *= 2
        return lists[0]
        
